<a href="https://colab.research.google.com/github/altn714/ESAA/blob/main/%EC%B7%A8%ED%95%A9_%EB%AF%B8%EB%8B%88%ED%94%8C%EC%A0%9D2_%EB%AF%BC%EA%B2%BD%2C%ED%98%84%EC%84%9C%EC%A0%84%EC%B2%98%EB%A6%AC%EB%B0%98%EC%98%81.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. 데이터셋 불러오기

In [75]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [76]:
import numpy as np
import pandas as pd 

In [77]:
train = pd.read_csv("/content/train 지역변수 추가.csv")
test = pd.read_csv("/content/test 지역변수 추가.csv")
sb = pd.read_csv("/content/submission_sample.csv")

In [78]:
test.head(3)

inst_id  OC     sido  sgg    openDate  bedCount          instkind  \
0        2 NaN  incheon  139  19981125.0     300.0  general_hospital   
1        5 NaN     jeju  149  20160309.0      44.0          hospital   
2        6 NaN  jeonnam  103  19890427.0     276.0  general_hospital   

       revenue1    salescost1          sga1  ...  liquidLiabilities2  \
0  6.682486e+10  0.000000e+00  6.565709e+10  ...        5.068443e+10   
1  3.495758e+10  0.000000e+00  3.259270e+10  ...        4.209828e+10   
2  2.326031e+10  2.542571e+09  2.308749e+10  ...        2.777589e+10   

     shortLoan2  NCLiabilities2     longLoan2     netAsset2      surplus2  \
0  3.714334e+10    4.720000e+09  4.690000e+09  1.608540e+10  8.944587e+09   
1  2.420000e+10    2.521009e+10  1.830000e+10  3.789135e+09  0.000000e+00   
2  2.182278e+10    0.000000e+00  0.000000e+00  0.000000e+00  1.638540e+10   

   employee1  employee2  ownerChange  hospital_count  
0      693.0      693.0         same             1.0  
1      379.0      371.0         same             1.0  
2        NaN        NaN          NaN             3.0  

[3 rows x 59 columns]

In [79]:
train.head(3)

inst_id    OC       sido  sgg  openDate  bedCount          instkind  \
0        1  open  choongnam   73  20071228     175.0  nursing_hospital   
1        3  open  gyeongnam   32  19970401     410.0  general_hospital   
2        4  open   gyeonggi   89  20161228     468.0  nursing_hospital   

       revenue1   salescost1          sga1  ...  liquidLiabilities2  \
0  4.217530e+09          0.0  3.961135e+09  ...         222876855.0   
1           NaN          NaN           NaN  ...                 NaN   
2  1.004522e+09  515483669.0  4.472197e+08  ...                 0.0   

   shortLoan2  NCLiabilities2    longLoan2     netAsset2      surplus2  \
0         0.0     536116887.0  390000000.0  2.619290e+09  1.271224e+09   
1         NaN             NaN          NaN           NaN           NaN   
2         0.0             0.0          0.0  0.000000e+00  0.000000e+00   

   employee1  employee2  ownerChange  hospital_count  
0       62.0       64.0         same            14.0  
1      801.0      813.0         same             5.0  
2      234.0        1.0         same            17.0  

[3 rows x 59 columns]

In [80]:
train.describe()

inst_id         sgg      openDate    bedCount      revenue1  \
count  301.000000  301.000000  3.010000e+02  296.000000  2.930000e+02   
mean   219.056478   81.039867  2.005013e+07  145.709459  1.288175e+10   
std    121.234869   50.969714  8.893815e+04  118.923890  2.043543e+10   
min      1.000000    1.000000  1.978012e+07    0.000000  0.000000e+00   
25%    112.000000   37.000000  2.001102e+07   52.750000  3.252112e+09   
50%    230.000000   75.000000  2.007113e+07  136.500000  5.524218e+09   
75%    321.000000  123.000000  2.011102e+07  193.000000  1.274839e+10   
max    428.000000  178.000000  2.017061e+07  656.000000  1.510000e+11   

         salescost1          sga1       salary1          noi1          noe1  \
count  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02   
mean   2.014903e+09  1.033244e+10  5.654115e+09  2.696151e+08  5.115870e+08   
std    7.460271e+09  1.493886e+10  8.083343e+09  8.023701e+08  1.060379e+09   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  2.758201e+09  1.626053e+09  8.217133e+06  8.013395e+07   
50%    2.104105e+08  4.684074e+09  2.659892e+09  4.363764e+07  1.831965e+08   
75%    9.105278e+08  1.077876e+10  6.363400e+09  2.050331e+08  4.203330e+08   
max    9.850332e+10  1.030000e+11  6.403559e+10  9.144171e+09  8.686380e+09   

       ...         debt2  liquidLiabilities2    shortLoan2  NCLiabilities2  \
count  ...  2.930000e+02        2.930000e+02  2.930000e+02    2.930000e+02   
mean   ...  8.146026e+09        3.860584e+09  1.510050e+09    4.471247e+09   
std    ...  1.255800e+10        6.797242e+09  2.953412e+09    7.659580e+09   
min    ...  0.000000e+00        0.000000e+00  0.000000e+00    0.000000e+00   
25%    ...  1.283220e+09        2.855741e+08  0.000000e+00    2.557878e+07   
50%    ...  3.784553e+09        1.454050e+09  8.542549e+07    1.969747e+09   
75%    ...  8.465053e+09        4.364714e+09  1.567967e+09    4.905441e+09   
max    ...  8.508858e+10        6.846878e+10  1.759375e+10    5.150388e+10   

          longLoan2     netAsset2      surplus2    employee1    employee2  \
count  2.930000e+02  2.930000e+02  2.930000e+02   291.000000   288.000000   
mean   2.709979e+09  5.273919e+09  9.786279e+08   142.546392   134.326389   
std    4.564001e+09  1.081259e+10  4.688798e+09   160.191073   151.061786   
min    0.000000e+00 -2.014780e+10 -2.781507e+09     0.000000     0.000000   
25%    0.000000e+00  1.017573e+09  0.000000e+00    53.500000    53.750000   
50%    1.100000e+09  2.894970e+09  0.000000e+00    80.000000    79.000000   
75%    3.360000e+09  5.370285e+09  1.014447e+08   181.500000   170.000000   
max    3.256147e+10  1.250000e+11  6.852730e+10  1200.000000  1200.000000   

       hospital_count  
count      299.000000  
mean        10.230769  
std          6.221061  
min          1.000000  
25%          5.000000  
50%         10.000000  
75%         15.500000  
max         20.000000  

[8 rows x 55 columns]

### # 변수설명
inst_id - 각 파일에서의 병원 고유 번호   
OC – 영업/폐업 분류, 2018년 폐업은 2017년 폐업으로 간주함    
sido – 병원의 광역 지역 정보    
sgg – 병원의 시군구 자료     
openDate – 병원 설립일    
bedCount - 병원이 갖추고 있는 병상의 수       
instkind – 병원, 의원, 요양병원, 한의원, 종합병원 등 병원의 종류    
·        종합병원 : 입원환자 100명 이상 수용 가능    
·        병원 : 입원 환자 30명 이상 100명 미만 수용 가능     
·        의원 : 입원 환자 30명 이하 수용 가능      
·        한방 병원(한의원) : 침술과 한약으로 치료하는 의료 기관.    
revenue1 – 매출액 / salescost1 – 매출원가      
sga1 - 판매비와 관리비 / salary1 – 급여   
noi1 – 영업외수익 / noe1 – 영업외비용      
Interest1 – 이자비용 / ctax1 – 법인세비용     
Profit1 – 당기순이익 / liquidAsset1 – 유동자산      
quickAsset1 – 당좌자산 / receivableS1 - 미수금(단기)       
inventoryAsset1 – 재고자산 / nonCAsset1 – 비유동자산     
tanAsset1 – 유형자산 / OnonCAsset1 - 기타 비유동자산       
receivableL1 – 장기미수금 / debt1 – 부채총계      
liquidLiabilities1 – 유동부채 / shortLoan1 – 단기차입금    
NCLiabilities1 – 비유동부채 / longLoan1 – 장기차입금    
netAsset1 – 순자산총계 / surplus1 – 이익잉여금    
employee1 – 고용한 총 직원의 수     
ownerChange – 대표자의 변동 

# 2. EDA 및 전처리

In [81]:
train.isnull().sum()

inst_id                0
OC                     0
sido                   0
sgg                    0
openDate               0
bedCount               5
instkind               1
revenue1               8
salescost1             8
sga1                   8
salary1                8
noi1                   8
noe1                   8
interest1              8
ctax1                  8
profit1                8
liquidAsset1           8
quickAsset1            8
receivableS1           8
inventoryAsset1        8
nonCAsset1             8
tanAsset1              8
OnonCAsset1            8
receivableL1           8
debt1                  8
liquidLiabilities1     8
shortLoan1             8
NCLiabilities1         8
longLoan1              8
netAsset1              8
surplus1               8
revenue2               8
salescost2             8
sga2                   8
salary2                8
noi2                   8
noe2                   8
interest2              8
ctax2                  8
profit2                8


In [82]:
test.isnull().sum()

inst_id                 0
OC                    127
sido                    0
sgg                     0
openDate                1
bedCount                8
instkind                2
revenue1                2
salescost1              2
sga1                    2
salary1                 2
noi1                    2
noe1                    2
interest1               2
ctax1                   2
profit1                 2
liquidAsset1            2
quickAsset1             2
receivableS1            2
inventoryAsset1         2
nonCAsset1              2
tanAsset1               2
OnonCAsset1             2
receivableL1            2
debt1                   2
liquidLiabilities1      2
shortLoan1              2
NCLiabilities1          2
longLoan1               2
netAsset1               2
surplus1                2
revenue2                2
salescost2              2
sga2                    2
salary2                 2
noi2                    2
noe2                    2
interest2               2
ctax2       

### (1) 레이블 데이터 : 인코딩

In [83]:
from sklearn.preprocessing import LabelEncoder

# OC : 'close': 0, 'open': 1 처리해주기
print(train['OC'].value_counts()) #인코딩전 확인
encoder1 = LabelEncoder()
train['OC']=encoder1.fit_transform(train['OC'])
print(train['OC'].value_counts()) #인코딩후 확인

# 테스트데이터셋은 OC가 모두 NaN이므로 레이블인코딩 처리x

open      286
 close     15
Name: OC, dtype: int64
1    286
0     15
Name: OC, dtype: int64


### (2) 피처 데이터 : 결측치 처리 및 인코딩

#### (2)-1. ints_id, sgg

In [84]:
# inst_id : 인덱스 중복이므로 열 버리기
# sgg : sido와 정보가 겹치므로 열 버리기
train.drop(['inst_id','sgg'],axis=1,inplace=True)
test.drop(['inst_id','sgg'],axis=1,inplace=True)

#### (2)-2. 문자열 피처

In [85]:
## 문자열 피처에서 결측치가 존재할 경우, 'Not_sure'로 대체
# 문자열 피처중 instkind, ownerChange 에서 결측치 존재

train['instkind'] = train['instkind'].fillna('Not_sure')
test['instkind'] = test['instkind'].fillna('Not_sure')

train['ownerChange'] = train['ownerChange'].fillna('Not_sure')
test['ownerChange'] = test['ownerChange'].fillna('Not_sure')

In [86]:
## 문자열 피처 인코딩 처리

# sido는 레이블인코딩 (sido의 개별관측치가 많아 원핫인코딩을 하게 되면 변수개수가 너무많아지기 때문.)
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
train['sido'] = encoder.fit_transform(train['sido'])
test['sido'] = encoder.fit_transform(test['sido'])

# instkind, ownerChange는 원핫인코딩
train = pd.get_dummies(train)
test = pd.get_dummies(test)

In [87]:
## 확인
train.head(3)
# sido는 숫자형변수로, instkind, ownerChange는 더미변수로 잘 바뀜.

OC  sido  openDate  bedCount      revenue1   salescost1          sga1  \
0   1     2  20071228     175.0  4.217530e+09          0.0  3.961135e+09   
1   1     9  19970401     410.0           NaN          NaN           NaN   
2   1     8  20161228     468.0  1.004522e+09  515483669.0  4.472197e+08   

        salary1        noi1        noe1  ...  instkind_clinic  \
0  2.033835e+09  15652441.0  15236236.0  ...                0   
1           NaN         NaN         NaN  ...                0   
2  2.964023e+08     76156.0     30000.0  ...                0   

   instkind_dental_clinic  instkind_general_hospital  instkind_hospital  \
0                       0                          0                  0   
1                       0                          1                  0   
2                       0                          0                  0   

   instkind_nursing_hospital  instkind_traditional_clinic  \
0                          1                            0   
1                          0                            0   
2                          1                            0   

   instkind_traditional_hospital  ownerChange_Not_sure  ownerChange_change  \
0                              0                     0                   0   
1                              0                     0                   0   
2                              0                     0                   0   

   ownerChange_same  
0                 1  
1                 1  
2                 1  

[3 rows x 66 columns]

#### (2)-3. 숫자형 피처중 금융변수

In [88]:
## 숫자형 피처중 금융변수에서 결측치(NaN) 존재

# open&같은행번호에서 공통적으로 8개의 결측치가 나타난다. 
print(train[train['revenue1'].isnull()]['OC'])
print(train[train['shortLoan2'].isnull()]['OC'])

# 같은행번호에서 공통적으로 2개의 결측치가 나타난다.
print(test[test['revenue1'].isnull()])
print(test[test['shortLoan2'].isnull()])

1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
    OC  sido    openDate  bedCount  revenue1  salescost1  sga1  salary1  noi1  \
15 NaN     0  19950330.0     226.0       NaN         NaN   NaN      NaN   NaN   
56 NaN    10  20121218.0     539.0       NaN         NaN   NaN      NaN   NaN   

    noe1  ...  instkind_Not_sure  instkind_clinic  instkind_general_hospital  \
15   NaN  ...                  0                0                          1   
56   NaN  ...                  0                0                          0   

    instkind_hospital  instkind_nursing_hospital  instkind_traditional_clinic  \
15                  0                          0                            0   
56                  0                          1                            0   

    instkind_traditional_hospital  ownerChange_Not_sure  ownerChange_change  \
15

In [89]:
#NAN인 관측치 처리(dropna)
fin_list=['revenue1', 'salescost1', 'sga1', 'salary1', 'noi1', 'noe1',
       'interest1', 'ctax1', 'profit1', 'liquidAsset1', 'quickAsset1',
       'receivableS1', 'inventoryAsset1', 'nonCAsset1', 'tanAsset1',
       'OnonCAsset1', 'receivableL1', 'debt1', 'liquidLiabilities1',
       'shortLoan1', 'NCLiabilities1', 'longLoan1', 'netAsset1', 'surplus1',
       'revenue2', 'salescost2', 'sga2', 'salary2', 'noi2', 'noe2',
       'interest2', 'ctax2', 'profit2', 'liquidAsset2', 'quickAsset2',
       'receivableS2', 'inventoryAsset2', 'nonCAsset2', 'tanAsset2',
       'OnonCAsset2', 'receivableL2', 'debt2', 'liquidLiabilities2',
       'shortLoan2', 'NCLiabilities2', 'longLoan2', 'netAsset2', 'surplus2']
train.dropna(subset=fin_list, axis=0,inplace=True)
test.dropna(subset=fin_list, axis=0,inplace=True)
#0인 관측치 처리
train[fin_list]=train[fin_list].replace(0 ,-9999)
test[fin_list]=test[fin_list].replace(0 ,-9999)

In [92]:
## 확인
train.describe()
test.describe()
# 최소값이 -9999로 나타난다.

OC        sido      openDate    bedCount      revenue1    salescost1  \
count  0.0  125.000000  1.240000e+02  117.000000  1.250000e+02  1.250000e+02   
mean   NaN    6.192000  2.005109e+07  169.205128  1.824446e+10  3.973317e+09   
std    NaN    4.142027  8.797709e+04  122.321336  3.133530e+10  1.419948e+10   
min    NaN    0.000000  1.976041e+07    0.000000 -9.999000e+03 -9.999000e+03   
25%    NaN    3.000000  2.000088e+07   94.000000  3.235889e+09 -9.999000e+03   
50%    NaN    6.000000  2.007107e+07  154.000000  6.637116e+09  1.896923e+08   
75%    NaN   10.000000  2.011345e+07  212.000000  1.704039e+10  1.192329e+09   
max    NaN   14.000000  2.017053e+07  771.000000  1.810000e+11  1.160000e+11   

               sga1       salary1          noi1          noe1  ...  \
count  1.250000e+02  1.250000e+02  1.250000e+02  1.250000e+02  ...   
mean   1.340410e+10  7.415083e+09  5.175509e+08  8.704298e+08  ...   
std    2.054481e+10  1.210043e+10  1.193139e+09  1.804462e+09  ...   
min   -9.999000e+03 -9.999000e+03 -9.999000e+03 -9.999000e+03  ...   
25%    3.083860e+09  1.666742e+09  1.641094e+07  8.365552e+07  ...   
50%    5.716502e+09  3.263054e+09  5.319171e+07  2.315190e+08  ...   
75%    1.445243e+10  7.198103e+09  3.714592e+08  6.633503e+08  ...   
max    1.270000e+11  8.239561e+10  7.302719e+09  1.149680e+10  ...   

       instkind_Not_sure  instkind_clinic  instkind_general_hospital  \
count          125.00000       125.000000                 125.000000   
mean             0.01600         0.032000                   0.144000   
std              0.12598         0.176708                   0.352503   
min              0.00000         0.000000                   0.000000   
25%              0.00000         0.000000                   0.000000   
50%              0.00000         0.000000                   0.000000   
75%              0.00000         0.000000                   0.000000   
max              1.00000         1.000000                   1.000000   

       instkind_hospital  instkind_nursing_hospital  \
count         125.000000                 125.000000   
mean            0.288000                   0.496000   
std             0.454653                   0.501996   
min             0.000000                   0.000000   
25%             0.000000                   0.000000   
50%             0.000000                   0.000000   
75%             1.000000                   1.000000   
max             1.000000                   1.000000   

       instkind_traditional_clinic  instkind_traditional_hospital  \
count                    125.00000                     125.000000   
mean                       0.01600                       0.008000   
std                        0.12598                       0.089443   
min                        0.00000                       0.000000   
25%                        0.00000                       0.000000   
50%                        0.00000                       0.000000   
75%                        0.00000                       0.000000   
max                        1.00000                       1.000000   

       ownerChange_Not_sure  ownerChange_change  ownerChange_same  
count            125.000000          125.000000         125.00000  
mean               0.120000            0.088000           0.79200  
std                0.326269            0.284435           0.40751  
min                0.000000            0.000000           0.00000  
25%                0.000000            0.000000           1.00000  
50%                0.000000            0.000000           1.00000  
75%                0.000000            0.000000           1.00000  
max                1.000000            1.000000           1.00000  

[8 rows x 65 columns]

#### (2)-4. bedcount 
instkind별 평균값을 bedcount의 결측값으로 처리

# 은서 전처리 반영전

In [29]:
train.groupby('instkind')['bedCount'].median()
train['bedCount'] = train['bedCount'].fillna(train.groupby('instkind')['bedCount'].transform('median')) 

train['bedCount'].isnull().sum()

KeyError: ignored

In [31]:
test.groupby('instkind')['bedCount'].median()
test['bedCount'] = test['bedCount'].fillna(test.groupby('instkind')['bedCount'].transform('median')) 
test['bedCount'].isnull().sum()
print(test['bedCount'])

KeyError: ignored

**test 파일에서 instkind별 bedCount 중앙값 결측치 처리 -> 123행 결측치 처리 안됨. (이유: traditional_hospital의 bedCount 행 하나밖에 없음.)**

In [ ]:
#결측치 행 추출
test[test['bedCount'].isnull()]

#### (2)-5. employee1,2 결측치 중앙값 대체

In [ ]:
#employee1 결측치 중앙값 대체 
train['employee1'].median()
train.loc[train['employee1'].isna(),'employee1'] = train['employee1'].median()
train['employee1'].isna().sum()

test['employee1'].median()
test.loc[test['employee1'].isna(),'employee1'] = test['employee1'].median()
test['employee1'].isna().sum()

In [ ]:
#employee2 결측치 중앙값 대체 
train['employee2'].median()
train.loc[train['employee2'].isna(),'employee2'] = train['employee2'].median()
train['employee2'].isna().sum()

test['employee2'].median()
test.loc[test['employee2'].isna(),'employee2'] = test['employee2'].median()
test['employee2'].isna().sum()

# 3. 파생변수 만들기

### (1) 금융변수

In [93]:
## 성장성
# 매출액 증가율 = 당기매출액/ 전기매출액*100 -100
train['revenueGrowth']=(train['revenue1']/train['revenue2'])*100-100
# 총자산 증가율 = 당기총자산/ 전기총자산*100 -100
train['netAssetGrowth']=(train['netAsset1']+train['debt1'])/(train['netAsset2']+train['debt2'])*100 -100

## 수익성
# 자기자본순이익률(ROE)=당기순이익/자기자본*100
train['ROE']=(train['profit1']/train['netAsset1'])*100
# 매출순이익률 = 당기순이익/매출액*100
train['ROS'] = train.profit1/train.revenue1*100
# 배당성향 = 배당금/당기순이익
train['dividendPayoutRatio'] = (train.profit1-train.surplus1)/train.profit1
# 이자보상비율 = 영업이익/이자비용*100
# (영업손익은 매출액 등의 영업수익으로부터 매출원가ㆍ판매비와관리비 등의 영업비용을 차감한 개념, 영업 외 이익과 비용도 차감)
train['interestCoverageRatio'] = (train.revenue1-train.salescost1-train.sga1-train.noi1-train.noe1)/train.interest1*100

## 안정성
# 유동비율 = 유동자산/유동부채*100 (바람직: 200%이상)
train['liquidRatio']=(train['liquidAsset1']/train['liquidLiabilities1'])*100
# 당좌비율 = 당좌자산/유동부채*100 (바람직: 100%이상)
train['quickRatio']=(train['quickAsset1']/train['liquidLiabilities1'])*100
# 부채비율 = 부채/자본*100
train['debtRatio']=(train['debt1']/(train['debt1']+train['netAsset1']))*100
# 자기자본비율 = 자기자본/총자본*100
train['netAssetRatio']=train.netAsset1/(train.debt1+train.netAsset1)*100

## 활동성
# 재고자산회전율 = 매출원가/평균재고자산
train['inventoryTurnover']=train.salescost1/((train.inventoryAsset1+train.inventoryAsset2)/2)

In [94]:
## 성장성
# 매출액 증가율 = 당기매출액/ 전기매출액*100 -100
test['revenueGrowth']=(test['revenue1']/test['revenue2'])*100-100
# 총자산 증가율 = 당기총자산/ 전기총자산*100 -100
test['netAssetGrowth']=(test['netAsset1']+test['debt1'])/(test['netAsset2']+test['debt2'])*100 -100

## 수익성
# 자기자본순이익률(ROE)=당기순이익/자기자본*100
test['ROE']=(test['profit1']/test['netAsset1'])*100
# 매출순이익률 = 당기순이익/매출액*100
test['ROS'] = test.profit1/test.revenue1*100
# 배당성향 = 배당금/당기순이익
test['dividendPayoutRatio'] = (test.profit1-test.surplus1)/test.profit1
# 이자보상비율 = 영업이익/이자비용*100
# (영업손익은 매출액 등의 영업수익으로부터 매출원가ㆍ판매비와관리비 등의 영업비용을 차감한 개념, 영업 외 이익과 비용도 차감)
test['interestCoverageRatio'] = (test.revenue1-test.salescost1-test.sga1-test.noi1-test.noe1)/test.interest1*100

## 안정성
# 유동비율 = 유동자산/유동부채*100 (바람직: 200%이상)
test['liquidRatio']=(test['liquidAsset1']/test['liquidLiabilities1'])*100
# 당좌비율 = 당좌자산/유동부채*100 (바람직: 100%이상)
test['quickRatio']=(test['quickAsset1']/test['liquidLiabilities1'])*100
# 부채비율 = 부채/자본*100
test['debtRatio']=(test['debt1']/(test['debt1']+test['netAsset1']))*100
# 자기자본비율 = 자기자본/총자본*100
test['netAssetRatio']=test.netAsset1/(test.debt1+test.netAsset1)*100

## 활동성
# 재고자산회전율 = 매출원가/평균재고자산
test['inventoryTurnover']=test.salescost1/((test.inventoryAsset1+test.inventoryAsset2)/2)

In [96]:
## 확인
fin_Ratio_list = ['revenueGrowth', 'netAssetGrowth',
    'ROE', 'ROS', 'dividendPayoutRatio', 'interestCoverageRatio',
    'liquidRatio', 'quickRatio', 'debtRatio','netAssetRatio',
    'inventoryTurnover']
train[fin_Ratio_list].describe()

fin_Ratio_list = ['revenueGrowth', 'netAssetGrowth',
    'ROE', 'ROS', 'dividendPayoutRatio', 'interestCoverageRatio',
    'liquidRatio', 'quickRatio', 'debtRatio','netAssetRatio',
    'inventoryTurnover']
test[fin_Ratio_list].describe()
# min값에 0이 없고, max값에 inf가 없다.

revenueGrowth  netAssetGrowth           ROE           ROS  \
count   1.250000e+02    1.250000e+02  1.250000e+02  1.250000e+02   
mean   -3.370253e+06   -5.693214e+06  6.498744e+05 -2.232017e+05   
std     1.829472e+07    2.212131e+07  3.579168e+06  2.495400e+06   
min    -1.351832e+08   -1.404061e+08 -3.411951e+06 -2.789942e+07   
25%    -3.135894e+00   -3.412206e+00  7.488287e-01  1.098195e-01   
50%     2.059830e+00    1.250984e+00  5.012549e+00  1.950740e+00   
75%     8.952727e+00    7.579308e+00  1.058053e+01  4.674831e+00   
max     9.362249e+03    1.498910e+02  2.501632e+07  9.991707e+01   

       dividendPayoutRatio  interestCoverageRatio  liquidRatio   quickRatio  \
count           125.000000           1.250000e+02   125.000000   125.000000   
mean             -5.588830           5.157842e+05   154.130059   149.951394   
std              27.685080           6.221739e+06   263.975141   260.520579   
min            -255.104334          -6.325280e+06     3.026015     3.026015   
25%              -3.335616          -5.368377e+02    46.516218    44.726958   
50%               0.999992           1.079300e+01    85.002275    81.073031   
75%               1.000062           1.332663e+02   145.171720   141.404901   
max              18.544894           6.912322e+07  1843.149250  1843.149250   

        debtRatio  netAssetRatio  inventoryTurnover  
count  125.000000     125.000000         125.000000  
mean    59.814847      40.185153       -4116.045756  
std     22.571632      22.571632       22000.901454  
min      2.391772     -34.901667     -172634.687569  
25%     49.160430      25.266250          -0.000007  
50%     59.451007      40.548993           4.837668  
75%     74.733750      50.839570          20.370486  
max    134.901667      97.608228        3019.942354

# 미수 전처리 반영전

### (2) 중소병원/대형병원

In [112]:
train.loc[train.bedCount>300,'hospital_size']=1
train.loc[train.bedCount<=300,'hospital_size']=0

### (3) 지역 내 병원종류수

In [117]:
#  이건 모델 돌리고나서 만들기

# 4. 피처 정규화

* 표준화 : 정규분포를 가진 값으로 변환 
            (x-mean)/sd
* 정규화 : 개별 데이터의 크기를 모두 똑같은 단위로 변경
            (x-min)/(max-min)

In [114]:
#금융변수 정규화
from sklearn.preprocessing import MinMaxScaler
train.loc[:,'revenue1':'surplus2'] = MinMaxScaler().fit_transform(train.loc[:,'revenue1':'surplus2'])
train.loc[:,'revenueGrowth':'inventoryTurnover'] = MinMaxScaler().fit_transform(train.loc[:,'revenueGrowth':'inventoryTurnover'])
test.loc[:,'revenue1':'surplus2'] = MinMaxScaler().fit_transform(test.loc[:,'revenue1':'surplus2'])
test.loc[:,'revenueGrowth':'inventoryTurnover'] = MinMaxScaler().fit_transform(test.loc[:,'revenueGrowth':'inventoryTurnover'])

In [115]:
#정규화 여부 확인
train.loc[:,'sido':'ownerChange_same']
train.loc[:,'revenueGrowth':'inventoryTurnover']

revenueGrowth  netAssetGrowth       ROE       ROS  dividendPayoutRatio  \
0         0.999993        0.999999  0.477187  0.004800             0.655805   
2         0.973021        0.988457  0.477187  0.004800             0.657435   
3         0.999993        0.999999  0.477187  0.004800             0.655087   
4         0.999993        0.999999  0.477188  0.004800             0.630024   
5         0.999993        0.999999  0.482359  0.004800             0.676878   
..             ...             ...       ...       ...                  ...   
296       0.999993        0.999999  0.477187  0.004800             0.656855   
297       0.819638        0.999999  0.658702  0.004800             0.657435   
298       0.999993        0.999999  0.477187  0.020080             0.657435   
299       0.999993        0.999999  0.684212  0.004798             0.657435   
300       0.999993        0.999999  0.477187  0.004800             0.691220   

     interestCoverageRatio  liquidRatio  quickRatio  debtRatio  netAssetRatio  \
0                 0.403513     0.999816    0.999814   0.036052       0.963948   
2                 0.400781     0.999808    0.999805   0.043800       0.956200   
3                 0.403502     0.999797    0.999795   0.140315       0.859685   
4                 0.403502     0.999796    0.999794   0.185888       0.814112   
5                 0.403502     0.999797    0.999795   0.093108       0.906892   
..                     ...          ...         ...        ...            ...   
296               0.403504     0.999803    0.999801   0.110484       0.889516   
297               0.403494     0.999800    0.999798   0.093108       0.906892   
298               0.403501     0.999803    0.999801   0.123317       0.876683   
299               0.403500     0.999798    0.999796   0.186213       0.813787   
300               0.403502     0.999796    0.999794   0.092492       0.907508   

     inventoryTurnover  
0             0.999609  
2             0.999660  
3             0.999609  
4             0.999609  
5             0.999667  
..                 ...  
296           0.999660  
297           0.233749  
298           0.999609  
299           0.999616  
300           0.999626  

[293 rows x 11 columns]

In [124]:
#전체 데이터 확인
pd.set_option('display.max_columns', None) # 전체 열 보기

In [127]:
train.describe()
#금융변수의 최솟값은 0 최댓값은 1로 정규화가 잘 되었다고 볼 수 있다. 

OC        sido      openDate    bedCount    revenue1  \
count  293.000000  293.000000  2.930000e+02  288.000000  293.000000   
mean     0.948805    7.098976  2.005059e+07  143.663194    0.085310   
std      0.220771    4.203769  8.854337e+04  118.062972    0.135334   
min      0.000000    0.000000  1.978012e+07    0.000000    0.000000   
25%      1.000000    3.000000  2.001122e+07   52.750000    0.021537   
50%      1.000000    8.000000  2.007120e+07  134.500000    0.036584   
75%      1.000000   10.000000  2.011102e+07  186.750000    0.084427   
max      1.000000   15.000000  2.017061e+07  656.000000    1.000000   

       salescost1        sga1     salary1        noi1        noe1   interest1  \
count  293.000000  293.000000  293.000000  293.000000  293.000000  293.000000   
mean     0.020455    0.100315    0.088297    0.029486    0.058896    0.065777   
std      0.075736    0.145037    0.126232    0.087747    0.122074    0.096381   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.026779    0.025393    0.000900    0.009226    0.017834   
50%      0.002136    0.045477    0.041538    0.004773    0.021091    0.038622   
75%      0.009244    0.104648    0.099373    0.022423    0.048391    0.078917   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

            ctax1     profit1  liquidAsset1  quickAsset1  receivableS1  \
count  293.000000  293.000000    293.000000   293.000000    293.000000   
mean     0.027939    0.403123      0.082615     0.081710      0.049501   
std      0.088509    0.093219      0.146840     0.145809      0.103269   
min      0.000000    0.000000      0.000000     0.000000      0.000000   
25%      0.000000    0.385679      0.016818     0.016287      0.000000   
50%      0.000368    0.392537      0.035431     0.035266      0.015318   
75%      0.015794    0.409931      0.078935     0.078338      0.049251   
max      1.000000    1.000000      1.000000     1.000000      1.000000   

       inventoryAsset1  nonCAsset1   tanAsset1  OnonCAsset1  receivableL1  \
count       293.000000  293.000000  293.000000   293.000000    293.000000   
mean          0.055839    0.073120    0.067526     0.033778      0.003413   
std           0.123404    0.109648    0.100931     0.092803      0.058421   
min           0.000000    0.000000    0.000000     0.000000      0.000000   
25%           0.002025    0.018590    0.017309     0.000097      0.000000   
50%           0.011652    0.039108    0.035573     0.003553      0.000000   
75%           0.050894    0.083161    0.077917     0.020211      0.000000   
max           1.000000    1.000000    1.000000     1.000000      1.000000   

            debt1  liquidLiabilities1  shortLoan1  NCLiabilities1   longLoan1  \
count  293.000000          293.000000  293.000000      293.000000  293.000000   
mean     0.101103            0.110554    0.088420        0.083310    0.064354   
std      0.148345            0.164430    0.160649        0.142637    0.114226   
min      0.000000            0.000000    0.000000        0.000000    0.000000   
25%      0.022528            0.012979    0.000000        0.007789    0.000000   
50%      0.051417            0.044567    0.011545        0.037024    0.027642   
75%      0.108861            0.137536    0.107605        0.089090    0.080902   
max      1.000000            1.000000    1.000000        1.000000    1.000000   

        netAsset1    surplus1    revenue2  salescost2        sga2     salary2  \
count  293.000000  293.000000  293.000000  293.000000  293.000000  293.000000   
mean     0.164165    0.030536    0.085468    0.021276    0.093552    0.081973   
std      0.079527    0.060244    0.138321    0.077894    0.137486    0.120865   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.134557    0.021290    0.019937    0.000000    0.022904    0.021997   
50%      0.145921    0.021290    0.036535    0.002002    0.042388   

In [116]:
#정규화 여부 확인
test.loc[:,'sido':'ownerChange_same']
test.loc[:,'revenueGrowth':'inventoryTurnover']

revenueGrowth  netAssetGrowth       ROE       ROS  dividendPayoutRatio  \
0         0.999931        0.999999  0.120020  0.999996             0.870989   
1         0.999931        0.999999  0.120020  0.999996             0.851400   
2         0.999931        0.999999  0.875945  0.999996             0.963752   
3         0.999931        0.999999  0.120020  0.000000             0.935885   
4         0.999931        0.999999  0.120020  0.999997             0.919586   
..             ...             ...       ...       ...                  ...   
122       0.999931        0.999999  0.120020  0.999997             0.935885   
123       0.999931        0.999999  0.120020  0.999996             0.422633   
124       0.999931        0.999999  0.057667  0.999996             0.935886   
125       0.999931        0.999999  0.120024  0.999997             0.935885   
126       0.999931        0.999999  0.000000  0.999997             0.935886   

     interestCoverageRatio  liquidRatio  quickRatio  debtRatio  netAssetRatio  \
0                 0.083830     0.024309    0.022780   0.672394       0.327606   
1                 0.083836     0.005920    0.005289   0.599565       0.400435   
2                 0.083827     0.022648    0.020125   0.359281       0.640719   
3                 1.000000     0.016395    0.014823   0.269306       0.730694   
4                 0.083836     0.015396    0.013852   0.498536       0.501464   
..                     ...          ...         ...        ...            ...   
122               0.083840     0.075660    0.075201   0.362716       0.637284   
123               0.083879     0.104740    0.104740   0.109247       0.890753   
124               0.083836     0.052700    0.052700   0.359281       0.640719   
125               0.083838     0.058168    0.050721   0.713626       0.286374   
126               0.083837     0.052700    0.052700   0.359281       0.640719   

     inventoryTurnover  
0             0.982807  
1             0.982807  
2             0.982821  
3             0.982807  
4             0.982807  
..                 ...  
122           0.982862  
123           0.982813  
124           0.000000  
125           0.983003  
126           0.317384  

[125 rows x 11 columns]

In [128]:
test.describe()
#금융변수의 최솟값은 0 최댓값은 1로 정규화가 잘 되었다고 볼 수 있다. 

OC        sido      openDate    bedCount    revenue1  salescost1  \
count  0.0  125.000000  1.240000e+02  117.000000  125.000000  125.000000   
mean   NaN    6.192000  2.005109e+07  169.205128    0.100798    0.034253   
std    NaN    4.142027  8.797709e+04  122.321336    0.173123    0.122409   
min    NaN    0.000000  1.976041e+07    0.000000    0.000000    0.000000   
25%    NaN    3.000000  2.000088e+07   94.000000    0.017878    0.000000   
50%    NaN    6.000000  2.007107e+07  154.000000    0.036669    0.001635   
75%    NaN   10.000000  2.011345e+07  212.000000    0.094146    0.010279   
max    NaN   14.000000  2.017053e+07  771.000000    1.000000    1.000000   

             sga1     salary1        noi1        noe1   interest1       ctax1  \
count  125.000000  125.000000  125.000000  125.000000  125.000000  125.000000   
mean     0.105544    0.089994    0.070872    0.075711    0.123099    0.049759   
std      0.161770    0.146858    0.163383    0.156953    0.178249    0.133845   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.024282    0.020229    0.002249    0.007277    0.020437    0.000000   
50%      0.045012    0.039602    0.007285    0.020139    0.058589    0.003658   
75%      0.113799    0.087360    0.050867    0.057700    0.131493    0.025970   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

          profit1  liquidAsset1  quickAsset1  receivableS1  inventoryAsset1  \
count  125.000000    125.000000   125.000000    125.000000       125.000000   
mean     0.361722      0.100707     0.100817      0.040597         0.098088   
std      0.107036      0.171446     0.173263      0.103888         0.175211   
min      0.000000      0.000000     0.000000      0.000000         0.000000   
25%      0.339089      0.018708     0.017983      0.000058         0.000690   
50%      0.347103      0.039535     0.039810      0.009813         0.023025   
75%      0.365174      0.097231     0.091185      0.044895         0.106946   
max      1.000000      1.000000     1.000000      1.000000         1.000000   

       nonCAsset1   tanAsset1  OnonCAsset1  receivableL1       debt1  \
count  125.000000  125.000000   125.000000    125.000000  125.000000   
mean     0.123473    0.127320     0.053719      0.008066    0.103702   
std      0.172095    0.180480     0.140662      0.089440    0.156321   
min      0.000000    0.000000     0.000000      0.000000    0.000000   
25%      0.035152    0.036210     0.000651      0.000000    0.016860   
50%      0.055828    0.056563     0.006857      0.000000    0.040919   
75%      0.118167    0.127766     0.031843      0.000000    0.114505   
max      1.000000    1.000000     1.000000      1.000000    1.000000   

       liquidLiabilities1  shortLoan1  NCLiabilities1   longLoan1   netAsset1  \
count          125.000000  125.000000      125.000000  125.000000  125.000000   
mean             0.099412    0.076922        0.079281    0.159105    0.198153   
std              0.166509    0.147871        0.127888    0.222881    0.192731   
min              0.000000    0.000000        0.000000    0.000000    0.000000   
25%              0.008374    0.000000        0.009967    0.015137    0.104341   
50%              0.033249    0.015500        0.034484    0.077199    0.130986   
75%              0.105951    0.073035        0.084403    0.191767    0.189977   
max              1.000000    1.000000        1.000000    1.000000    1.000000   

         surplus1    revenue2  salescost2        sga2     salary2        noi2  \
count  125.000000  125.000000  125.000000  125.000000  125.000000  125.000000   
mean     0.248473    0.098028    0.034040    0.099214    0.084051    0.067287   
std      0.130284    0.171531    0.121632    0.157195    0.141996    0.162093   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.202021    0.017464    0.000000    0.021837    0.017291    0.001612   
50%      0.20202

# 5. 모델 돌리기

## (1) RandomForest

## (2) XGBoost

## (3) LightGBM

In [ ]:
X_features = train.drop(columns = 'OC',axis=1)
y_labels = train['OC']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, test_size=0.2, random_state=156)

In [ ]:
# LightGBM의 파이썬 패키지인 lightgbm에서 LGBMClassifier 임포트
from lightgbm import LGBMClassifier
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# 앞서 XGBoost와 동일하게 n_estimators는 400설정
lgbm_wrapper = LGBMClassifier(n_estimators=400)

# LightGBM도 XGBoost와 동일하게 조기중단 수행가능
evals = [(X_test, y_test)]
lgbm_wrapper.fit(X_train, y_train,
                 early_stopping_rounds=100, eval_metric="logloss",
                 eval_set=evals, verbose=True)
preds = lgbm_wrapper.predict(X_test)
pred_proba = lgbm_wrapper.predict_proba(X_test)[:, 1]

In [ ]:
type(train['ownerChange'])

In [ ]:
train['ownerChange'].value_counts()

In [34]:
train[train['ownerChange'].isnull()]

KeyError: ignored

In [ ]:
train['ownerChange'].isna().sum()